### DISCUSSION TITLE
RFC: customizing buildtype

### DISCUSSION THREAD

**Discussion by bruchar1** [mesonbuild/meson#12720] [Ideas]
I had an idea about an improvement for `buildtype`, and I'd like to get some feedback about it.

My idea would be to add a new function to `meson.options` file:

`buildtype('name', debug: true, optimization: '1', ...)`

where keywords would be any existing option (builtin, or user defined).

That function would allow three things:
- redefine existing buildtypes, but with different values for `debug` or `optimization`
- define new buildtypes for the project
- define default values for any option, for the specified buildtype.

For instance, one could define the default `b_ndebug` option per buildtype like this:
```
buildtype('release', b_ndebug: true)
buildtype('plain', b_ndebug: true)
buildtype('debugoptimized', b_ndebug: false)
...
```

Same for `b_vscrt`, or for any other option (builtin or user).

Having this would lessen the need for special option values like `'if-release'` for `b_ndebug`, or `'from_buildtype'` and `'static_from_buildtype'` for `b_vscrt`, and would allow other options to depend on the buildtype, without special treatment.

The order of precedence would be:
- hardcoded default value
- meson.options file
- project `default_options`
- `buildtype` options
- machine files
- command line

What do you think of this?

**Comment by eli-schwartz** [comment]
> redefine existing buildtypes, but with different values for debug or optimization

CMake allows redefining existing build types and it's an endless source of trouble. IMO this should never ever ever be allowed. If a built-in buildtype is irreparably broken for a given project, simply define a new build type and use that instead. Check for its use in meson.build and raise an error() if users use it.

**Reply by bruchar1** [reply]
You probably have more experience than me about this point... What I'm thinking about are things like: "I want a 'release' build with `optimization: '2'`", "I want a 'plain' build with/without debug info...", "I want a 'debugoptimized' build with "optimization: '3'".

Redefining those things could be convenient for interoperability with other tools, like using conan for fetching dependencies and generating machine file (like I do in my project). 

On the other hand, I think my proposition would still be very usable (and useful) even if it disallows buildtype redefinition.
